### PR TITLE
Use BindingHostKey annotation to detect scheduled pods

### LINKED ISSUE #388: Pre-scheduled pods not supported
Kubernetes allows to set the NodeName value of a pod explicitly, bypassing the scheduler. The pod is started by the kubelet on the respective node.

### PR DESCRIPTION
**Pull request by sttts**
Before NodeName in the pod spec was used. Hence, pods with a fixed, pre-set
NodeName were never scheduled, leading e.g. to a failing e2e intra-pod test.

Fixes mesosphere/kubernetes-mesos#388

### CODE REVIEW

**Comment by sttts** [comment]
Temporarily including commit of #11108 which conflicts with these changes otherwise.

File: contrib/mesos/pkg/scheduler/plugin.go [hunk 1]
```diff
@@ -430,7 +430,7 @@ func (q *queuer) Run(done <-chan struct{}) {
 			}
 
 			pod := p.(*Pod)
-			if pod.Spec.NodeName != "" {
+			if _, ok := pod.ObjectMeta.Annotations[annotation.BindingHostKey]; ok {
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/plugin.go]: the k8s stock scheduler uses pod.Spec.NodeName. why are we changing?

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go [hunk 2]
```diff
@@ -213,6 +213,13 @@ func (t *T) AcceptOffer(offer *mesos.Offer) bool {
 	if offer == nil {
 		return false
 	}
+
+	// possibly check NodeName
+	if t.Pod.Spec.NodeName != "" && *offer.Hostname != t.Pod.Spec.NodeName {
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: `offer.GetHostname()` is marginally safer

File: contrib/mesos/pkg/scheduler/plugin.go (hunk 1 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/plugin.go]: Because in upstream the kubelet will notice any task with a pre-set NodeName and launch it. In the k8sm case we have to create a Mesos task, launch an executor and then pass the pod to the kubelet. In other words: if we only filter for NodeName, we miss those pre-scheduled tasks in the k8sm scheduler.

Comment (sttts) [inline contrib/mesos/pkg/scheduler/plugin.go]: Compare https://github.com/mesosphere/kubernetes-mesos/issues/388

Comment (jdef) [inline contrib/mesos/pkg/scheduler/plugin.go]: perhaps we can encapsulate this field & map lookup in a private package func? (and also test for Annotations != nil before the k/v lookup) with some documentation for the func that explains why we're not using pod.Spec.NodeName. otherwise it's pretty confusing when comparing schedulers.

File: contrib/mesos/pkg/executor/executor.go [hunk 3]
```diff
@@ -463,17 +449,66 @@ func (k *KubernetesExecutor) launchTask(driver bindings.ExecutorDriver, taskId s
...
+		if pod.Spec.NodeName != k.hostname {
+			log.Errorf("Pod %v/%v bound to another node: %v", pod.Namespace, pod.Name, pod.Spec.NodeName)
+			deleteTask()
+			k.sendStatus(driver, newStatus(mutil.NewTaskID(taskId), mesos.TaskState_TASK_FAILED,
+				messages.WrongSlaveFailure))
+			return
+		}
+
+		// post annotations update to apiserver
+		patch := struct {
```
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: is using anonymous/inline structs the k8s blessed way of applying patches?

File: contrib/mesos/pkg/scheduler/plugin.go (hunk 1 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/plugin.go]: Agree. This needs a clean-up and documentation.

File: contrib/mesos/pkg/executor/executor.go [hunk 4]
```diff
@@ -463,17 +449,66 @@ func (k *KubernetesExecutor) launchTask(driver bindings.ExecutorDriver, taskId s
...
+		err := k.client.Post().Namespace(api.NamespaceValue(ctx)).Resource("bindings").Body(binding).Do().Error()
+		if err != nil {
+			deleteTask()
+			k.sendStatus(driver, newStatus(mutil.NewTaskID(taskId), mesos.TaskState_TASK_FAILED,
+				messages.CreateBindingFailure))
+			return
+		}
+	} else {
+		// we have a pod which is already assigned to a node, hopefully to this slave
+		if pod.Spec.NodeName != k.hostname {
```
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: tracking the slave hostname in k.hostname and then comparing it to pod.Spec.NodeName smells funny. can we instead compare pod.Spec.NodeName to pod.ObjectMeta.Annotations[annotation.BindingHostKey]? we could then drop k.hostname from the struct.

File: contrib/mesos/pkg/executor/executor.go [hunk 5]
```diff
@@ -129,6 +129,7 @@ type KubernetesExecutor struct {
 	staticPodsConfig     []byte
 	staticPodsConfigPath string
 	initialRegComplete   chan struct{}
+	hostname             string
```
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: would be nice to avoid tracking this. see later comments.

File: contrib/mesos/pkg/executor/executor.go (hunk 3 above)
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: Patching is not used much throughout the code base. But where it is used, it's done with inline structs, yes. Git grep for '.patch('.

File: contrib/mesos/pkg/executor/executor.go (hunk 4 above)
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: Good idea. We have that annotation anyway here.

File: contrib/mesos/pkg/scheduler/plugin.go (hunk 1 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/plugin.go]: done.

File: contrib/mesos/docs/architecture.md [hunk 6]
```diff
@@ -18,6 +18,25 @@ The executor launches the pod/task, which registers the bound pod with the kubel
...
+
+   - The offer is *accepted*,
+   - the pod is *annotated* with a number of annotation, especially `k8s.mesosphere.io/bindingHost`
+   - the pod is *launched* on a Mesos slave.
+
+   The existence of the `bindingHost` annotation tells the k8sm-scheduler that this pod has been launched. If it is not set, the pod is considered *new*.
+
+2. The Mesos slave receives the task launch event and starts (if not running yet) the k8sm-executor (possibly via the km hyperkube binary). Then:
+
+    - The k8sm-scheduler *binds* the tasks to the node via the apiserver, which means that the `NodeName` field is set by the apiserver.
```
Comment (jdef) [inline contrib/mesos/docs/architecture.md]: actually the executor does the binding after it receives the launchTask from mesos

File: contrib/mesos/pkg/scheduler/plugin.go (hunk 1 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/plugin.go]: nice

File: contrib/mesos/docs/architecture.md (hunk 6 above)
Comment (sttts) [inline contrib/mesos/docs/architecture.md]: Of course, typo. Thanks.

**Comment by sttts** [comment]
@jdef ptal

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go [hunk 7]
```diff
@@ -213,6 +213,13 @@ func (t *T) AcceptOffer(offer *mesos.Offer) bool {
 	if offer == nil {
 		return false
 	}
+
+	// possibly check NodeName
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: nit: could we improve this comment? something along the lines of:
if the user has specified a target host make sure it's aligned with this offer

**Comment by jdef** [comment]
minor nit, otherwise lgtm

**Comment by sttts** [comment]
Squashed.

@jdef or @karlkfi Ready to be merged. Address all comments by James. PR only changes contrib/mesos.

**Comment by mikedanese** [comment]
@sttts needs rebase

**Comment by sttts** [comment]
@mikedanese rebased. ready to be merged.
